core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -25,34 +37,291 @@ type GCSFileTransfer struct {
-	ft.logger.Debug("GCSFileTransfer: Upload: uploading file", "path", task.Path)
+	ft.logger.Debug("GCSFileTransfer: Upload: uploading file", "path", task.PathOrPrefix)
 
 	return fmt.Errorf("not implemented yet")
 }
 
-// Download downloads a file from the server
+// Download downloads a file from the server.
 func (ft *GCSFileTransfer) Download(task *ReferenceArtifactDownloadTask) error {
-	ft.logger.Debug("GCSFileTransfer: Download: downloading file", "path", task.Path, "ref", task.Reference)
+	ft.logger.Debug(
+		"GCSFileTransfer: Download: downloading file",
+		"path", task.PathOrPrefix,
+		"ref", task.Reference,
+	)
 
-	return fmt.Errorf("not implemented yet")
+	// Parse the reference path to get the scheme, bucket, and object
+	bucketName, objectPathPrefix, err := parseReference(task.Reference)
+	if err != nil {
+		return formatDownloadError("error parsing reference", err)
+	}
+	bucket := ft.client.Bucket(bucketName)
+
+	var objectNames []string
+
+	if task.HasSingleFile() {
+		objectNames = []string{objectPathPrefix}
+	} else {
+		objectNames, err = ft.listObjectNamesWithPrefix(bucket, objectPathPrefix)
+		if err != nil {
+			return formatDownloadError(
+				fmt.Sprintf(
+					"error getting objects in bucket %s under prefix %s",
+					bucketName, objectPathPrefix,
+				),
+				err,
+			)
+		}
+	}
+
+	return ft.downloadFiles(bucket, task, objectNames)
+}
+
+// listObjectNamesWithPrefix returns a list of names of all the objects
+// with a specified prefix in the given bucket.
+func (ft *GCSFileTransfer) listObjectNamesWithPrefix(
+	bucket *storage.BucketHandle,
+	prefix string,
+) ([]string, error) {
+	var objectNames []string
+	query := &storage.Query{Prefix: prefix}
+	it := bucket.Objects(ft.ctx, query)
+	for {
+		objAttrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		objectNames = append(objectNames, objAttrs.Name)
+	}
+	return objectNames, nil
+}
+
+// downloadFiles concurrently gets and downloads all objects with the specified names
+func (ft *GCSFileTransfer) downloadFiles(
+	bucket *storage.BucketHandle,
+	task *ReferenceArtifactDownloadTask,
+	objectNames []string,
+) error {
+	// Parse the reference path to get the objectPathPrefix, for constructing the download path
+	_, objectPathPrefix, err := parseReference(task.Reference)
+	if err != nil {
+		return formatDownloadError("error parsing reference", err)
+	}
+
+	g := new(errgroup.Group)
+	g.SetLimit(maxWorkers)
+	for _, name := range objectNames {
+		objectName := name // for closure in the goroutine
+		g.Go(func() error {
+			object, objAttrs, err := ft.getObjectAndAttrs(bucket, task, objectName)
+			if err != nil {
+				if errors.Is(err, ErrObjectIsDirectory) {
+					ft.logger.Debug(
+						"GCSFileTransfer: Download: skipping reference because it seems to be a folder",
+						"reference", task.Reference,
+						"object", objectName,
+					)
+					return nil
+				}
```

Comment:
Can this still happen? Doesn't this code list objects by prefix when given a directory?

Is it even possible for a task to have Digest != Reference but refer to a folder?

Reply:
yeah if you create a folder in gcloud with the UI, as in by clicking the "new folder" button there, you'll actually create an object for that folder and it'll be returned from the `listObjects` functions in the gcloud sdk (in python and go), but you won't be able to get its attributes or download it. we patched this in python so we no longer will create entries for objects corresponding to folders when uploading, but to ensure backwards compatibility I think we need to keep this